### ISSUE TITLE
Allow the geometry of a polygon layer to be "inverted"

### ISSUE THREAD

**Issue by peterqliu** [mapbox/mapbox-gl-js#993]
**tl;dr** a layout option to render the areas polygons _aren't_

I'd like to make a flood map where water levels rise from low to high. If we imagined this happening on an island:

![screen shot 2015-02-11 at 4 41 06 pm](https://cloud.githubusercontent.com/assets/4551508/6159624/e386e150-b20c-11e4-89b4-ccf5dff141da.png)

The water would creep from outside in, and I would simulate this by adding successive doughnut-shaped water layers with a hole in the middle, where the land is too high to be flooded (this hole would get progressively smaller as water level rises).

However, Mapbox Terrain provides elevations only as **closed polygons**, which are the doughnut holes in our pastry analogy. I would like a layout option to switch from `doughnut hole -> doughnut`

cc @jfirebaugh @ansis

**Comment by jfirebaugh** [comment]
Can't you do this with a "water" background layer at the bottom and then hide the land pancakes from the bottom up (low elevation to high elevation)?

**Comment by peterqliu** [comment]
@jfirebaugh i can if all the land is one color. When we add landcovers and roads, that'll sit on top of the water-- I think i'll actually need a water layer **on top** of everything, to simulate the raised water level

**Comment by ansis** [comment]
Can you handle this on the data side by inverting the layer before creating the vector tiles?

**Comment by peterqliu** [comment]
@ansis per chat with @MateoV, it would be pretty heavy to do this for the whole planet. But I also think this would be useful elsewhere: imagine `render-inverse` on the `#water` layer to select all landmasses-- this would free us to put land on top of water, enabling us to do drop shadows of land onto water

![screen shot 2015-02-11 at 5 18 37 pm](https://cloud.githubusercontent.com/assets/4551508/6160064/b246cbcc-b212-11e4-85aa-19c18f66ba22.png)

**Comment by ljbade** [comment]
This sounds hard now that the stencil buffer has been dropped.

**Comment by mourner** [comment]
> @ansis per chat with @MateoV, it would be pretty heavy to do this for the whole planet.

@peterqliu actually it would be very easy to do on the JS side after we're done with https://github.com/mapbox/mapnik-vector-tile/issues/59.

So lets say we have many multipolygons in the input data that look like this:

```js
var multipolygons = [
  [outerRingA, hole1A, hole2A, ...], // poly A
  [outherRingB, hole1B, hole2B, ...], // poly B
  ...
];
```

An inverse of these then we'll look like this:

```js
var inverseMultopolygons = [
  [bigSquare, outerRingA, outerRingB, ...], // outer rings become holes
  [hole1A],
  [hole2A],
  [hole1B],
  [hole2B]
]

**Comment by mourner** [comment]
@springmeyer yet another use case for a defined ring order in VT (in addition to tessellation)

**Comment by springmeyer** [comment]
:+1: @mourner

**Comment by jfirebaugh** [comment]
I think this feature is far down enough on our priority list that, in the interest of keeping the issue count manageable, we need to close it. You may be able to implement the effect outside of the library by reversing the geometry rings manually as suggested by @mourner.

**Comment by Stophface** [comment]
You can use turf's mask function for that
https://www.npmjs.com/package/@turf/mask
Simply create a "cover polygon" with the extend of your (Map)View (or the whole world) and use your Polygon. Turf even creates the "cover polygon" for you, if you do not provide one.
https://github.com/Turfjs/turf/blob/master/packages/turf-mask/index.js#L58

**Comment by indus** [comment]
@Stophface how would you do this for the whole planet z0-14?

**Comment by Stophface** [comment]
@indus That question does not make much sense, or I understand it wrongly. If you have vector data, this data will be displayed the same for each zoom level.

**Comment by indus** [comment]
@Stophface
For a simple geometry, your suggestion is fine. In case you don't want it to do on the client but in preprocessing I would suggest mapshaper. 
Maybe I was unclear what case I asked for. I was talking specifically about masking `water` over various zoom levels. The complete geometry of "Natural Earth 10m" is really huge, but that is what is needed to mask seamless without any overlaps on z14. The information is already in the 'water' layer but it would need to be inverted to be useful.

**Comment by Stophface** [comment]
@indus Again, if you have Vector File this file will always be displayed the same, no matter on what zoom level you are. 

Your problem is: "My input file (geojson?) is really large. How can I display large geojson on a map?"

That has nothing to do with masking. One approach could (!) be simplifying your input data according to the zoom levels. You might end up with 14 different files, one file for each zoom level. You also could clip your large input file to the mapview the user is currently looking at (this will most likely not run seemlessly). There are various ways how to handle large geojson on a map.

**Comment by markusjohnsson** [comment]
@Stophface No, I think @indus is talking about vector tiles which has different level of details at different zoom levels. I had the exact same problem. I wanted to represent _land_ which would be the inverse of _water_ . I wanted to use the data provided by mapbox, only style it differently, but it was impossible. 

I put a couple of days into implementing it in mapbox-gl, had a promising test case, but some serious bugs. Never got any help or interest from the team.. and then they changed the license too and I got discouraged.

**Comment by indus** [comment]
@markusjohnsson thats exacly my problem. Maybe we can get some traction here: [#1206](https://github.com/openmaptiles/openmaptiles/issues/1206)
To make it more easy to create a VT `land` layer

**Comment by markusjohnsson** [comment]
@indus yeah it seems like there is such an elegant solution to the problem. I'm not working with that project any more so I won't spend a lot of time on it, sorry. Good luck to you though 🙂
